core/camel-util/src/main/java/org/apache/camel/util/ReflectionHelper.java

```diff
@@ -177,7 +177,7 @@ public static Method findMethod(Class<?> clazz, String name, Class<?>... paramTy
         return null;
     }
 
-    public static void setField(Field f, Object instance, Object value) {
+    public synchronized static void setField(Field f, Object instance, Object value) {
```

Comment:
It is not exactly what I had in mind, I was more thinking of removing the code that restores the old value of the flag this way we don't need any synchronization mechanism but we can wait for comments from others